### DISCUSSION TITLE
type errors in actions of machine when using createModel, model.createMachine, and model.assign

### DISCUSSION THREAD

**Discussion by bylethviva** [statelyai/xstate#2477] [Q&A] [answered]
I'm starting to experiment with Models and I'm running into some unexpected type errors with one of the ways you can create actions. Specifically if I pass in an options object when using `model.createMachine`, similar to [this section](https://xstate.js.org/docs/guides/models.html#modeling-events) of the docs, I get a type error on my action. I looked into the types for this section and `model.assign` is returning `AssignAction<TContext, ExtractEvent<TEvent, TEventType>>`, while the `machineOptions.actions` is supposed to have the following signature, `Record<string, ActionObject<TContext, TEvent> | ActionFunction<TContext, TEvent>>`. 

Am I doing something wrong here? The other two methods of creating an action [here](https://xstate.js.org/docs/guides/models.html#creating-a-machine-from-a-model) and [here](https://xstate.js.org/docs/guides/models.html#narrowing-assign-event-types), both work as expected.

[Here's a codesandbox](https://codesandbox.io/s/nervous-lake-f7im8?file=/src/exportsMachine.ts) that has all three cases.

**Comment by davidkpiano** [comment, marked as answer]
Please see the warning here: https://xstate.js.org/docs/guides/models.html#narrowing-assign-event-types

> Assign actions with narrowed event types cannot be placed inside the `actions: {...}` property of machine options in `createMachine(configuration, options)`. This is because actions in `options.actions` should be assumed to potentially receive any event, even if the machine configuration suggests otherwise.

**Reply by bylethviva** [reply]
I saw that before and clearly didn't internalize what it meant 😅  Thank you so much for your quick response!

**Reply by bylethviva** [reply]
@davidkpiano is it possible to add multiple assign actions with narrowed event types in the actions array of a state node? e.g.

```typescript
      error: {
        on: {
          exportRetry: {
            target: 'exporting',
            actions: [updateExportRetries, setErrorMsg],
            cond: 'hasExportRetriesLeft',
          },
        },
      },
```

Currently when I do this, there are no errors in VS Code, but if I run `npx tsc` in the terminal it finds an error.
